src/main/java/net/neoforged/neoforge/registries/DeferredHolder.java

```diff
@@ -0,0 +1,268 @@
+        this.bind(false);
+    }
+
+    /**
+     * Gets the object stored by this DeferredHolder, if this holder {@linkplain #isPresent() is present}.
+     *
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    @SuppressWarnings("unchecked")
+    @Override
+    public T value() {
+        bind(true);
+        if (this.holder == null) {
+            throw new NullPointerException("Trying to access unbound value: " + this.key);
+        }
+
+        return (T) this.holder.value();
+    }
+
+    /**
+     * Gets the object stored by this DeferredHolder, if this holder {@linkplain #isPresent() is present}.
+     *
+     * @throws IllegalStateException If the backing registry is unavailable.
+     * @throws NullPointerException  If the underlying Holder has not been populated (the target object is not registered).
+     */
+    @Override
+    public T get() {
+        return this.value();
+    }
+
+    /**
+     * @return An optional containing the target object, if {@link #isPresent()}, otherwise {@linkplain Optional#empty() an empty optional}.
+     */
+    public Optional<T> asOptional() {
+        return isPresent() ? Optional.of(value()) : Optional.empty();
+    }
+
+    /**
+     * @return The registry that this DeferredHolder is pointing at, or {@code null} if it doesn't exist.
+     */
+    @Nullable
+    @SuppressWarnings("unchecked")
+    protected Registry<R> getRegistry() {
+        return (Registry<R>) BuiltInRegistries.REGISTRY.get(this.key.registry());
+    }
+
+    /**
+     * Binds this DeferredHolder to the underlying registry and target object.
+     *
+     * <p>Has no effect if already bound.
+     *
+     * @param throwOnMissingRegistry If true, an exception will be thrown if the registry is absent.
+     * @throws IllegalStateException If throwOnMissingRegistry is true and the backing registry is unavailable.
+     */
+    protected void bind(boolean throwOnMissingRegistry) {
+        if (this.holder != null) return;
+
+        Registry<R> registry = getRegistry();
+        if (registry != null) {
+            this.holder = registry.getHolder(this.key).orElse(null);
+        } else if (throwOnMissingRegistry) {
+            throw new IllegalStateException("Registry not present for " + this + ": " + this.key.registry());
+        }
+    }
+
+    /**
+     * @return The ID of the object pointed to by this DeferredHolder.
+     */
+    public ResourceLocation getId() {
+        return this.key.location();
+    }
+
+    /**
+     * @return The ResourceKey of the object pointed to by this DeferredHolder.
+     */
+    public ResourceKey<R> getKey() {
+        return this.key;
+    }
+
+    /**
+     * @return True if this DeferredHolder has been bound, and {@link #get()} or {@link #value()} may be called.
+     */
+    public boolean isPresent() {
```

Comment:
Why do we have both this function and `isBound`? What's the point? We immediately bind holders when they are added to the registry.

Reply:
Removed in favor of `isBound()`.